### PR TITLE
Add DataSetAttributes

### PR DESCRIPTION
**Pull request by JevinJ**
### Overview
Part 3 of #517 split, waiting for #657

This is larger than I would like, but the tests require integration with the other classes. 

I had an issue where endianness gets swapped when converting cell arrays from numpy to vtk, causing test_meshio beam and uniform tests to fail. I was able to fix failing tests by swapping from big to little before conversion but I don't know if this is correct. (line 216-218) in DataSetAttributes.append()

This is pretty much done so I'll accept input while waiting for #657.

Major changes:
* Replaces cell_arrays, field_arrays, point_arrays implementations.
* Removes _ScalarsDict, PointScalarsDict, CellScalarsDict, FieldScalarsDict
* Moves t_coords to DataSetAttributes, mirroring vtk.
* Removed test_cell_array_non_contiguous(), test is invalid because the case is handled in append().
* (Possible regression) Removes default behavior when getting a row array through the private _row_array(), the key is required, however there were previously no public paths for the default behavior. 
* Modified pyvista_ndarray to work with current implementation, allows the underlying vtkArray to be accessed through pyvista_ndarray.
* Add hexbeam fixture.

### CODE REVIEW

**Comment by akaszynski** [comment]
Looks like we're almost there except for:
```python
ValueError: Here is a summary of the problems encountered when running the examples

Unexpected failing examples:
/home/vsts/work/1/s/examples/02-plot/labels.py failed leaving traceback:
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 460, in _memory_usage
    out = func()
  File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 442, in __call__
    exec(self.code, self.fake_main.__dict__)
  File "/home/vsts/work/1/s/examples/02-plot/labels.py", line 36, in <module>
    plotter.add_point_labels(poly, "My Labels", point_size=20, font_size=36)
  File "/home/vsts/work/1/s/pyvista/plotting/plotting.py", line 2995, in add_point_labels
    labels = points.point_arrays[labels].astype(str)
AttributeError: 'vtkCommonCorePython.vtkStringArray' object has no attribute 'astype'
```

**Comment by JevinJ** [comment]
@akaszynski Fixed.

File: pyvista/core/datasetattributes.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+    def __init__(self, vtkobject, dataset, association):
+        """Initialize DataSetAttributes."""
+        super().__init__(vtkobject=vtkobject)
+        self.dataset = dataset
+        self.association = association
+
+    def __getitem__(self, key):
+        """
+        Implement get behaviour of [] operator.
```
Comment (akaszynski) [inline pyvista/core/datasetattributes.py:38]: Let's be consistent with the docstring; first line shouldn't be empty.

Comment (JevinJ) [reply pyvista/core/datasetattributes.py]: Fixed.

File: pyvista/core/common.py [hunk 2]
```diff
@@ -189,22 +190,7 @@ def _field_array(self, name=None):
             Numpy array of scalars
 
         """
-        if name is None:
-            raise RuntimeError('Must specify an array to fetch.')
```
Comment (akaszynski) [inline pyvista/core/common.py:193]: Use `ValueError` here.  See:
https://docs.python.org/3/library/exceptions.html#ValueError
vs.
https://docs.python.org/3/library/exceptions.html#RuntimeWarning

**Comment by akaszynski** [comment]
This is great work!  I'd avoid force pushing as this can delete other people's history, though I understand that `git` can be frustrating.

Made some minor changes to the docs while I went through.  None were strictly necessary, but it helps me review the code if I'm "fixing" it.

**Comment by akaszynski** [comment]
@JevinJ, it's up to you when you want others to review it, but I'd say it's good to go.  This is the direction I'd like the code to move in.  Extend, don't replace existing functionality like `VTKObjectWrapper`, and divide up the code into smaller modules.  Moving 300+ out of `common.py` was great.

**Comment by akaszynski** [comment]
I'm seeing this come up in the CI tests:
```
tests/test_filters.py::test_cell_centers_composite PASSED                [ 30%]
##[error]Cmd.exe exited with code '-1073740940'.
```
Error -1073740940 is `STATUS_HEAP_CORRUPTION, which is fairly serious and this will be a tough bug to track down.  One of the runs it occured right after `test_cell_centers_composite`.  If you're running windows, can you see if you can reproduce this bug on your end?  It's not happening on my end for Linux.

**Comment by JevinJ** [comment]
So, as you might see from the automated tests, I'm getting intermittent segfaults from test_glyph. I'm having trouble getting any information about the cause or when it fails.

> 
> 
> I'm seeing this come up in the CI tests:
> 
> ```
> tests/test_filters.py::test_cell_centers_composite PASSED                [ 30%]
> ##[error]Cmd.exe exited with code '-1073740940'.
> ```
> 
> Error -1073740940 is `STATUS_HEAP_CORRUPTION, which is fairly serious and this will be a tough bug to track down. One of the runs it occured right after `test_cell_centers_composite`. If you're running windows, can you see if you can reproduce this bug on your end? It's not happening on my end for Linux.

I'm aware of this, it's an intermittent segfault in test_glyph and it appears to happen for all platforms, just didn't happen with linux this run. I managed to get this out of faulthandler last night:

> Windows fatal exception: code 0xc0000374

Current thread 0x00008574 (most recent call first):
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\vtk\util\numpy_support.py", line 101 in create_vtk_array
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\vtk\util\numpy_support.py", line 147 in numpy_to_vtk
  File "C:\Users\User\PycharmProjects\pyvista\pyvista\core\pointset.py", line 363 in _from_arrays
  File "C:\Users\User\PycharmProjects\pyvista\pyvista\core\pointset.py", line 156 in __init__
  File "C:\Users\User\PycharmProjects\pyvista\pyvista\core\filters.py", line 986 in glyph
  File "C:\Users\User\PycharmProjects\pyvista\tests\test_filters.py", line 383 in test_glyph
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\python.py", line 184 in pytest_pyfunc_call
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\python.py", line 1479 in runtest
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\runner.py", line 135 in pytest_runtest_call
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\runner.py", line 217 in <lambda>
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\runner.py", line 244 in from_call
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\runner.py", line 217 in call_runtest_hook
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\runner.py", line 186 in call_and_report
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\runner.py", line 100 in runtestprotocol
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\runner.py", line 85 in pytest_runtest_protocol
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\main.py", line 272 in pytest_runtestloop
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\main.py", line 247 in _main
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\main.py", line 191 in wrap_session
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\main.py", line 240 in pytest_cmdline_main
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\callers.py", line 187 in _multicall
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\pluggy\hooks.py", line 286 in __call__
  File "C:\Users\User\PycharmProjects\pyvista\venv\lib\site-packages\_pytest\config\__init__.py", line 125 in main
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2019.2.3\plugins\python-ce\helpers\pycharm\_jb_pytest_runner.py", line 42 in <module>

I'll likely need help for this one and I'd appreciate it if you could take a look.

**Comment by akaszynski** [comment]
Looking at this and we're going to have to refactor `pyvista_ndarray`.  It's not ok to be monkey patching:
```python
obj.VTKObject = vtk_array
```
I think we're getting caught up with references to deleted VTK objects, and we need to use a `weakref.ref`, but looking at this more makes me realize that we're doing somethings that are unnecessary.  For example, in `common.py`
```python
        vtk_data = pts.GetData()
        arr = vtk_to_numpy(vtk_data)
        return pyvista.pyvista_ndarray(arr, vtk_data)
```
Should just be:
```python
pyvista.pyvista_ndarray(pts.GetData())
```
Like in the other classes in `pyvista`, the wrapping should take place in the `__init__` method.

Also, there's other consequences to returning a `pyvista.pyvista_ndarray`.  For example:

```python
import pyvista
sphere = pyvista.Sphere()
pyvista.plot(sphere.points)
```
```
TypeError: Object type (<class 'pyvista.core.pyvista_ndarray.pyvista_ndarray'>) not supported for plotting in PyVista.
```
Which means the base class for `pyvista_ndarray` should be a `np.ndarray` no a `vtk.VTKArray`.  There's other issues as well.  I'll probably commit to this tonight.

**Comment by akaszynski** [comment]
@JevinJ, that's better, thanks!

**Comment by JevinJ** [comment]
pyvista_ndarray may certainly need to be changed, it's entirely a dataset_adapter.VTKArray with __new__ modified to store either objects which are VTKObjectWrappers or pyvista objects + some helpful classmethod constructors.

**Comment by akaszynski** [comment]
I'm going to submit a PR for `pyvista_ndarray`.  There's several code breaking changes to one of my private projects that I've just noticed.  If that gets accepted and merged, we can go from there.

**Comment by akaszynski** [comment]
@JevinJ, can we break this up a little more?  This is exactly the direction we need to go, but these memory bugs are really going to drive us nuts.

BTW, you've really put in some great work here and I don't want this to go away.  Our package has really suffered from code bloat and this really cleans it up.  Refactoring is really tough, and it seems that we need to work on our unit testing even more.

If we can break it up into smaller chunks, we'll be able to see which commit causes that heap memory error.  Does this sound good to you @JevinJ?

**Comment by banesullivan** [comment]
I want to reiterate @akaszynski's note: @JevinJ, we *so greatly appreciate* all of the work you are putting into PyVista lately! These are absolutely wonderful contributions and so desperately needed! I know our requests to break these changes up more might be irritating but we are finding ourselves a bit overwhelmed and want to make sure these high-quality changes are able to be included in an upcoming release.

Unfortunately, I think @akaszynski and I need to focus on improving PyVista's testing coverage and functional test cases before we can merge big changes like this and some of the other PRs - this will prevent us from overlooking how these changes affect downstream projects (especially the closed source ones).

I'll work hard this week to try to improve our testing to get us to that point and then put 100% into reviewing your PRs.

Also, @JevinJ, please feel free to log onto the PyVsita Slack workspace and chat with either @akaszynski, me, or the `#core-dev` channel

**Comment by JevinJ** [comment]
@banesullivan, @akaszynski One of the major differences in how adding arrays is handled here is that there are no references to numpy arrays stored like ScalarsDict did. The problem is losing reference to the original narray when using numpy_to_vtk will invalidate the vtk array, unless it's deep copied. It's more of an issue somewhere in append() than any specific commit made elsewhere.

This does not seem to occur if they are deep copied, after running the test with deep copy ~50 times it does not occur as opposed to every ~5 runs otherwise. Using numpyTovtkDataArray instead of pyvista.convert_array is showing promise also, but string arrays are not handled.

**Comment by JevinJ** [comment]
For anyone having a look at this I did find that calling DataSetAttributes.update() in test_glygh precipitates this issue, I'm tending to think the original arrays are get deallocated but I'm not sure why yet.

**Comment by JevinJ** [comment]
@pyvista/developers, @akaszynski, @banesullivan I believe I've fixed it, this is good for review now. 

I don't think the currently failed tests are an issue with this PR, at least for MacOS. The Windows error seems to be a file system error, but difficult for me to debug without being able to run plotting. 

The segfault I had previously seems to be caused by me assigning numpy arrays in update() directly without running them through np.array. I'd appreciate it if anyone knows why that would be an issue.

**Comment by akaszynski** [comment]
Since this is going to be a nightmare of a bug to track down, you might try implementing the PR in steps.  It's such a big PR, it's tough for me to see where we're introducing the bug.

**Comment by JevinJ** [comment]
@akaszynski is this good to merge? Mac on 3.6 still fails.

**Comment by JevinJ** [comment]
@pyvista/developers, @banesullivan, @akaszynski This is ready for review/approval(for real this time).

**Comment by akaszynski** [comment]
@JevinJ,

Looks like we're almost good, but it seems we don't test `vtkStringArrays` in the core api and that's tripping us up in the documentation testing.

```python
poly = pv.PolyData(np.random.rand(10, 3))
poly["My Labels"] = ["Label {}".format(i) for i in range(poly.n_points)]
```

It looks like the current `pyvista_ndarray` isn't able to handle string arrays.  Could you add that functionality?

In this branch you would get:

```python
>>> poly["My Labels"]
pyvista_ndarray([(vtkCommonCorePython.vtkStringArray)0x7f2a342b6d00],
                dtype=object)
```

However, in master you get:

```python
>>> poly["My Labels"]
array([b'Label 0', b'Label 1', b'Label 2', b'Label 3', b'Label 4',
       b'Label 5', b'Label 6', b'Label 7', b'Label 8', b'Label 9'],
      dtype='|S7')
```

**Comment by akaszynski** [comment]
@JevinJ,

I recommend just running the failing example until you get this right, so you don't have to wait the 23 minutes for docbuild to complete with:

```
python examples/02-plot/labels.py
```

Also, would you mind adding in a test for string arrays within `tests/test_datasetattributes.py`?  The reason why we're not failing on the core API tests is because that edge case isn't being tested.

**Comment by JevinJ** [comment]
> Also, would you mind adding in a test for string arrays within tests/test_datasetattributes.py? The reason why we're not failing on the core API tests is because that edge case isn't being tested.

Done.

**Comment by akaszynski** [comment]
I've got a fix for this.  One second while I optimize.

**Comment by JevinJ** [comment]
Is |S(byte strings) desirable for the resulting array?

**Comment by akaszynski** [comment]
Actually, unicode is probably better.  Feel free to fix it.

**Comment by JevinJ** [comment]
I'm trying to figure out a case where ['0', ...] is converted to np.uint8.

**Comment by akaszynski** [comment]
Just force the datatype, otherwise numpy will assume it might be an int.

**Comment by JevinJ** [comment]
Ok, I got something working, in convert_string_array() instead of checking for ValueError, I'm checking if the array type is a string type.

With hypothesis instead of returning if the example is invalid, and you can't make a strategy do exactly what you want, you should use assume(). Doing so marks the example as invalid, so it doesn't get run again.

**Comment by akaszynski** [comment]
I see what you did there:
```python
@given(arr=arrays(dtype='U', shape=10))
```
Cool!  I'm starting to like `hypothesis`.

**Comment by JevinJ** [comment]
If there aren't any more issues this is ready.

**Comment by akaszynski** [comment]
I’ll check this out this morning.

**Comment by akaszynski** [comment]
@JevinJ, I'm merging this as our `qt_plotter` tests are flaky on MacOS.
